compiler+runtime/src/cpp/jank/read/lex.cpp

```diff
@@ -377,30 +402,88 @@ namespace jank::read
               auto const c(oc.unwrap());
               if(c == '.')
               {
-                if(contains_dot || !contains_leading_digit)
+                if(contains_dot || is_scientific || !contains_leading_digit)
                 {
                   ++pos;
                   return err(error{ token_start, pos, "invalid number" });
                 }
                 contains_dot = true;
               }
+              else if(c == 'e' || c == 'E')
+              {
+                if(is_scientific || !contains_leading_digit)
+                {
+                  ++pos;
+                  return err(error{ token_start, pos, "invalid number" });
+                }
+                is_scientific = true;
+                expecting_exponent = true;
+              }
+              else if(c == '+' || c == '-')
+              {
+                if(found_exponent_sign || !is_scientific || !expecting_exponent)
+                {
+                  ++pos;
+                  return err(error{ token_start, pos, "invalid number" });
+                }
+                found_exponent_sign = true;
+              }
+              else if(c == '/') {
+                require_space = false;
+                ++pos;
+                if(found_exponent_sign || is_scientific || expecting_exponent || contains_dot
+                  || found_slash_after_number) {
+                  return err(error{ token_start, pos, "invalid number" });
+                }
+                found_slash_after_number = true;
+                ++pos; // skip the '/' char and look for the denominator number.
+                const result<token, error> denominator = next();
+                if (denominator.is_ok() && denominator.expect_ok().kind == token_kind::integer) {
+                  const token& denominator_token = denominator.expect_ok();
+                  found_slash_after_number = false;
+                  return ok(token(token_start,
+                                 pos - token_start,
+                                 token_kind::ratio, {
+                                   .numerator = std::strtoll(file.data() + token_start, nullptr, 10),
+                                   .denominator = boost::get<native_integer>(denominator_token.data)}));
+
+                }
+                return err(error{ token_start, pos, "invalid number" });
```

Comment:
What's a more helpful error message we can put here? We know that we're expecting an integer denominator and we didn't get one.

Reply:
i changed it to "invalid ratio: expecting an integer denominator".